alt-tab-macos/logic/Application.swift

```diff
@@ -0,0 +1,99 @@
+        } else {
+            self.runningApplication.addObserver(self, forKeyPath: "isFinishedLaunching", options: [.new], context: nil)
+        }
+    }
+
+    private func addAndObserveWindows() {
+        axUiElement = AXUIElementCreateApplication(runningApplication.processIdentifier)
+        AXObserverCreate(runningApplication.processIdentifier, axObserverApplicationCallback, &axObserver)
+        observeAllWindows()
+    }
+
+    private func observeAllWindows() {
+        let windows = getActualWindows()
+        debugPrint("Adding app: " + (runningApplication.bundleIdentifier ?? "nil"), windows.map { $0.title() })
+        addWindows(windows)
+        observeEvents(windows)
+    }
+
+    func observeNewWindows() {
+        var newWindows = [AXUIElement]()
+        for window in getActualWindows() {
+            guard Windows.listRecentlyUsedFirst.firstIndexThatMatches(window) == nil else { continue }
+            newWindows.append(window)
+        }
+        addWindows(newWindows)
+    }
+
+    private func getActualWindows() -> [AXUIElement] {
+        return axUiElement!.windows()?.filter { $0.isActualWindow(runningApplication.isHidden) } ?? []
+    }
+
+    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey: Any]?, context: UnsafeMutableRawPointer?) {
+        guard let isFinishedLaunching = change![.newKey], isFinishedLaunching as! Bool else { return }
+        addAndObserveWindows()
+        runningApplication.removeObserver(self, forKeyPath: "isFinishedLaunching")
+    }
+
+    private func addWindows(_ windows: [AXUIElement]) {
+        Windows.listRecentlyUsedFirst.insert(contentsOf: windows.map { Window($0, self) }, at: 0)
+    }
+
+    private func observeEvents(_ windows: [AXUIElement]) {
+        guard let axObserver = axObserver else { return }
+        let selfPointer = UnsafeMutableRawPointer(Unmanaged.passUnretained(self).toOpaque())
+        AXObserverAddNotification(axObserver, axUiElement!, kAXApplicationActivatedNotification as CFString, selfPointer)
+        AXObserverAddNotification(axObserver, axUiElement!, kAXFocusedWindowChangedNotification as CFString, selfPointer)
+        AXObserverAddNotification(axObserver, axUiElement!, kAXWindowCreatedNotification as CFString, selfPointer)
+        AXObserverAddNotification(axObserver, axUiElement!, kAXApplicationHiddenNotification as CFString, selfPointer)
+        AXObserverAddNotification(axObserver, axUiElement!, kAXApplicationShownNotification as CFString, selfPointer)
+        // TODO: when using the global queue, sometimes it breaks the `runningApplications` observation somehow
+        // To reproduce, open Terminal, quit it, open, quit, repeat. After a while the .insertion and .removal events will stop happening
+//        DispatchQueue.global(qos: .userInteractive).async {
+            CFRunLoopAddSource(CFRunLoopGetCurrent(), AXObserverGetRunLoopSource(axObserver), .defaultMode)
+//            CFRunLoopRun()
```

Comment:
@koekeishiya this piece of code here, and the general GCD system are really a mystery to me. I spent a huge amount of time documenting myself on the topic, but it's such a mess of layer on top of layers of legacy APIs.

I used 2 dedicated `DispatchQueue` in v3, and a few `DispatchQueue.global.async` and `DispatchQueue.main.async` calls. However the fact that I had to comment the global dispatch here is where I'm hitting a wall.

If you have experience with GCD and the run loops, could you possibly take a look at my approach here?